src/app/util/attribute-storage.cpp

```diff
@@ -1379,3 +1418,30 @@ Optional<AttributeId> emberAfGetServerAttributeIdByIndex(EndpointId endpoint, Cl
     }
     return Optional<AttributeId>(clusterObj->attributes[attributeIndex].attributeId);
 }
+
+DataVersion * emberAfDataVersionStorage(chip::EndpointId endpointId, chip::ClusterId clusterId)
+{
+    uint16_t index = emberAfIndexFromEndpoint(endpointId);
+    if (index == 0xFFFF)
+    {
+        // Unknown endpoint.
+        return nullptr;
+    }
+    const EmberAfDefinedEndpoint & ep = emAfEndpoints[index];
+    if (!ep.dataVersions)
+    {
+        // No storage provided.
+        return nullptr;
+    }
+
+    // This does a second walk over endpoints to find the right one, but
+    // probably worth it to avoid duplicating code.
+    auto clusterIndex = emberAfClusterIndex(endpointId, clusterId, CLUSTER_MASK_SERVER);
+    if (clusterIndex == 0xFF)
+    {
+        // No such cluster on this endpoint.
+        return nullptr;
+    }
+
+    return ep.dataVersions + clusterIndex;
```

Comment:
The nature of this indexing is non-obvious. The comment about this structural requirement should be clearer:

```
// dataVersionStorage needs to point to a memory area that holds one DataVersion
// for every server cluster on this endpoint.  If nullptr is passed, the
```

^^^ points to memory of a flat sequence of data versions, one for each cluster on the endpoint, in numerical order of the clusters

The method used, of having a flat array, also forces us to never be able to provide a full code-based cluster handling (where we delegate to logic the current version getting), since we always need to follow the Ember flat model of data structures that are just flat in cluster order, as we code-generate them today. It seems that trying to backport this new feature in the "legacy" low-level ember approach is not necessary and adds constraints for future improvements.

Reply:
Do you have a concrete counter-proposal for where to put the data version for a cluster?  We've had a _lot_ of back-and-forth about this between me, @mrjerryjohns, and @yunhanw-google and this is the solution we settled on.

Do you have use cases for maintaining a data version outside the core IM implementation for 1.0?  Because we can absolutely re-evaluate this as we go on, but for the moment this is the simplest thing we can do that does not involve spec surgery, minimizes storage needed in the general case, and does not introduce too much additional complexity into the IM.

Reply:
>  The comment about this structural requirement should be clearer

Oh, and that requirement is not something the consumer needs to know about.  The only requirement is that the buffer is big enough; how we put the data versions in there is our business.  It might NOT be in numerical order of the clusters!  In fact, it's in the order the clusters come in in the endpoint type for the endpoint, which might not be numerical order, and for dynamic endpoints is up to the SDK consumer anyway.

But I will update the comments to make it clearer what the sizing situation is.  The use of Span will make it clear what the general alignment and whatnot requirements are.

Reply:
@tcarmelveilleux  just pull you to im dev slack channel for all the discussion on this topic, yes, this current soluation is what we finally settle on

Reply:
FWIW @tcarmelveilleux I agree with your points, and if I read between the tea leaves there, I think you're likely headed in the path of adding the data version lookup in the `AttributeAccessInterface` provider.

I think similar to how we achieve over-rides for attribute values, I could see us adding data-version over-ride look-up there as well. I think that could slot in nicely on-top of the work here.

That cannot by itself however, be sufficient.